### PR TITLE
Update kuberay mcad integration doc

### LINKED ISSUE #1327: Update MCAD/CodeFlare documentation to improve usability
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

The MCAD documentation for usage alongside KubeRay[1] is quite old and hard to follow. We should update this document to make it easier to follow and bring it up to date with the latest in the CodeFlare ecosystem.

[1] https://github.com/ray-project/kuberay/blob/master/docs/guidance/kuberay-with-MCAD.md

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by tedhtchang**
## Why are these changes needed?

Closes #1327 Improve the kuberay macad integration doc

## Related issue number
#1327 

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by tedhtchang** [comment]
This PR should be review together with the [PR](https://github.com/project-codeflare/multi-cluster-app-dispatcher/pull/617).
/cc @anishasthana @kevin85421

File: docs/guidance/kuberay-with-MCAD.md [hunk 1]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
 - Support for multiple Kubernetes clusters; dispatching jobs to any one of a number of Kubernetes clusters.
 
 
-In order to queue Ray cluster(s) and `gang dispatch` them when aggregated resources are available please refer to the setup [KubeRay-MCAD integration](https://github.com/IBM/multi-cluster-app-dispatcher/blob/quota-management/doc/usage/examples/kuberay/kuberay-mcad.md) with configuration files [here](https://github.com/IBM/multi-cluster-app-dispatcher/tree/quota-management/doc/usage/examples/kuberay/config).
+In order to queue Ray cluster(s) and `gang dispatch` them when aggregated resources are available please refer to the setup [KubeRay-MCAD integration](https://github.com/IBM/multi-cluster-app-dispatcher/blob/quota-management/doc/usage/examples/kuberay/kuberay-mcad.md) on a Kubernetes Cluster or an OpenShift Cluster.
```
Comment (anishasthana) [inline docs/guidance/kuberay-with-MCAD.md:17]: This url should be updated to refer to the CodeFlare repository instead of `IBM`

File: docs/guidance/kuberay-with-MCAD.md [hunk 2]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
...
+
+For OpenShift Cluster, the MCAD and KubeRay are already part of CodeFlare stack which provides a simple, user-friendly abstraction for scaling, queuing and resource management of distributed AI/ML and Python workloads. Please follow the Quick Start in the [Distributed Workloads](https://github.com/opendatahub-io/distributed-workloads) for installation.
```
Comment (anishasthana) [inline docs/guidance/kuberay-with-MCAD.md:19]:
```suggestion
On OpenShift,  MCAD and KubeRay are already part of the Open Data Hub Distributed Workload Stack. The stack provides a simple, user-friendly abstraction for scaling, queuing and resource management of distributed AI/ML and Python workloads. Please follow the Quick Start in the [Distributed Workloads](https://github.com/opendatahub-io/distributed-workloads) for installation.
```

File: docs/guidance/kuberay-with-MCAD.md [hunk 3]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
...
-Events:                          <none>
-```
-
-
-- As seen the second Ray cluster is queued with no pending pods created. 
-
-- Dispatching policy out of the box is FIFO which can be augmented as per user needs. The second cluster will be dispatched when additional aggregated resources are available in the cluster or the first AppWrapper Ray cluster is deleted.
+Let's create two Ray clusters with the AppWrapper custom resource on the same Kubernetes cluster.
+
+- Assuming you have installed all the pre-requisites mentioned in the [KubeRay-MCAD integration](https://github.com/IBM/multi-cluster-app-dispatcher/blob/quota-management/doc/usage/examples/kuberay/kuberay-mcad.md), we submit the first Ray cluster with the AppWrapper CR [aw-raycluster.yaml](https://github.com/IBM/multi-cluster-app-dispatcher/blob/quota-management/doc/usage/examples/kuberay/config/aw-raycluster.yaml):
```
Comment (anishasthana) [inline docs/guidance/kuberay-with-MCAD.md:26]: Fix IBM references to be CodeFlare

**Review by anishasthana — CHANGES_REQUESTED** [review]

File: docs/guidance/kuberay-with-MCAD.md [hunk 4]
```diff
@@ -5,7 +5,7 @@ The multi-cluster-app-dispatcher is a Kubernetes controller providing mechanisms
 ## Use case
 
 MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources are available in the cluster prior to actual pod creation in the Kubernetes cluster. It supports features such as:
-   
+
 - Integrates with upstream Kubernetes scheduling stack for features such co-scheduling, Packing on GPU dimension etc.
 - Ability to wrap any Kubernetes objects.
 - Increases control plane stability by JIT (Just-in Time) object creation.
```
Comment (kevin85421) [inline docs/guidance/kuberay-with-MCAD.md:11]: Out of curiosity, would you mind explaining this feature?

File: docs/guidance/kuberay-with-MCAD.md [hunk 5]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
...
-  Sender:                        before ScheduleNext - setHOL
-  State:                         Pending
-Events:                          <none>
-```
-
-
-- As seen the second Ray cluster is queued with no pending pods created. 
-
-- Dispatching policy out of the box is FIFO which can be augmented as per user needs. The second cluster will be dispatched when additional aggregated resources are available in the cluster or the first AppWrapper Ray cluster is deleted.
+Let's create two Ray clusters with the AppWrapper custom resource on the same Kubernetes cluster.
```
Comment (kevin85421) [inline docs/guidance/kuberay-with-MCAD.md:24]:
```suggestion
Let's create two RayCluster custom resources with the AppWrapper custom resource on the same Kubernetes cluster.
```

File: docs/guidance/kuberay-with-MCAD.md (hunk 3 above)
Comment (kevin85421) [inline docs/guidance/kuberay-with-MCAD.md:26]:
```suggestion
- Assuming you have installed all the pre-requisites mentioned in the [KubeRay-MCAD integration](https://github.com/IBM/multi-cluster-app-dispatcher/blob/quota-management/doc/usage/examples/kuberay/kuberay-mcad.md), we submit the first RayCluster with the AppWrapper CR [aw-raycluster.yaml](https://github.com/IBM/multi-cluster-app-dispatcher/blob/quota-management/doc/usage/examples/kuberay/config/aw-raycluster.yaml):
```

File: docs/guidance/kuberay-with-MCAD.md [hunk 6]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
...
+    Sender:                        before manageQueueJob - afterEtcdDispatching
+    State:                         Running
+  Events:                          <none>
+  (base) asmalvan@mcad-dev:~/mcad-kuberay$ kubectl get pod -n default
+  NAME                                           READY   STATUS    RESTARTS   AGE
+  raycluster-complete-head-9s4x5                 1/1     Running   0          47s
+  raycluster-complete-worker-small-group-4s6jv   1/1     Running   0          47s
+  ```
+
+- As seen the cluster is dispatched and pods are running. Let's submit another Ray cluster with AppWrapper CR and see it queued without creating pending pods using the command:
```
Comment (kevin85421) [inline docs/guidance/kuberay-with-MCAD.md:72]:
```suggestion
- As seen the cluster is dispatched and Pods are running. Let's submit another RayCluster with AppWrapper CR and see it queued without creating pending Pods using the command:
```

File: docs/guidance/kuberay-with-MCAD.md [hunk 7]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
...
+  ```bash
+  kubectl create -f https://raw.githubusercontent.com/project-codeflare/multi-cluster-app-dispatcher/quota-management/doc/usage/examples/kuberay/config/aw-raycluster-1.yaml
+
+  # check the raycluster-complete-1 appwrapper
+  kubectl describe  appwrapper raycluster-complete-1 -n default
+  ``````
```
Comment (kevin85421) [inline docs/guidance/kuberay-with-MCAD.md:78]: "``````" -> "```"

File: docs/guidance/kuberay-with-MCAD.md [hunk 8]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
...
+  kubectl create -f https://raw.githubusercontent.com/project-codeflare/multi-cluster-app-dispatcher/quota-management/doc/usage/examples/kuberay/config/aw-raycluster.yaml
+  ```
+  Check Appwrapper status
+  ```
+  kubectl describe  appwrapper raycluster-complete -n default
+  ```
+
+  Expect:
+  ```
+  Status:
```
Comment (kevin85421) [inline docs/guidance/kuberay-with-MCAD.md:38]: Which field users should focus on? I cannot understand the "expectation" here.

File: docs/guidance/kuberay-with-MCAD.md [hunk 9]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
...
+- As seen the cluster is dispatched and pods are running. Let's submit another Ray cluster with AppWrapper CR and see it queued without creating pending pods using the command:
+  ```bash
+  kubectl create -f https://raw.githubusercontent.com/project-codeflare/multi-cluster-app-dispatcher/quota-management/doc/usage/examples/kuberay/config/aw-raycluster-1.yaml
+
+  # check the raycluster-complete-1 appwrapper
+  kubectl describe  appwrapper raycluster-complete-1 -n default
+  ``````
+  Expect
+  ```
+  Status:
```
Comment (kevin85421) [inline docs/guidance/kuberay-with-MCAD.md:81]: Which fields should users focus on?

File: docs/guidance/kuberay-with-MCAD.md [hunk 10]
```diff
@@ -14,83 +14,101 @@ MCAD allows you to deploy Ray cluster with a guarantee that sufficient resources
...
+      Type:                        Backoff
+    Controllerfirsttimestamp:      2023-08-29T17:39:08.406399Z
+    Filterignore:                  true
+    Queuejobstate:                 Backoff
+    Sender:                        before [backoff] - Rejoining
+    State:                         Pending
+  Events:                          <none>
+  ```
+
+  As seen the second Ray cluster is queued with no pending pods created.
```
Comment (kevin85421) [inline docs/guidance/kuberay-with-MCAD.md:111]: Every user allocates different CPU/memory resources to their Kubernetes clusters. If a user possesses a high-end workstation, would the RayCluster still be queued?

Comment (tedhtchang) [reply docs/guidance/kuberay-with-MCAD.md:111]: No likely won't be queue. My cluster has only has 16 cpus with some overhead already allocated so the 2nd AW queued. If the high-end workstation has 16+ cpus, the 2nd AppWrapper may not go over total allocatable cpus.

File: docs/guidance/kuberay-with-MCAD.md (hunk 4 above)
Comment (tedhtchang) [reply docs/guidance/kuberay-with-MCAD.md:11]: @asm582 may explain this JIT better.
My understanding is mcad creating object only when there is enough resource. In other words, there will not be any pending pods so the control plane stability is improved.

Comment (tedhtchang) [reply docs/guidance/kuberay-with-MCAD.md:11]: My other thought would be, Mcad when paired the InstaScale operator can auto scale out enough k8s worker nodes to run a job and scales down afterwards; This kinda fits the Just in time concept of allocating right amount resources on the assembly line to complete a task; in order to reduce inventory and cost.

File: docs/guidance/kuberay-with-MCAD.md (hunk 10 above)
Comment (kevin85421) [reply docs/guidance/kuberay-with-MCAD.md]: We should provide a clear example so users can consistently reproduce the expected behavior across all environments.

Comment (kevin85421) [reply docs/guidance/kuberay-with-MCAD.md]: Could you also explain when MCAD creates the RayCluster CR (e.g., 5 CPUs)? Without this information, it's difficult for users to understand how MCAD operates.

**Review by asm582 — APPROVED** [review]
LGTM

File: docs/guidance/kuberay-with-MCAD.md (hunk 10 above)
Comment (tedhtchang) [reply docs/guidance/kuberay-with-MCAD.md]: @kevin85421 I have added more commits. Do you think these 2 comments have been addressed ?

**Review by anishasthana — APPROVED** [review]
/lgtm

File: docs/guidance/kuberay-with-MCAD.md (hunk 4 above)
Comment (anishasthana) [reply docs/guidance/kuberay-with-MCAD.md:11]: Yep that's about right! MCAD will not create the underlying Ray resources until there are enough resources to schedule the pod. If InstaScale is enabled, InstaScale will scale up new nodes on your cluster -> MCAD will schedule Ray resources onto new nodes -> Ray Cluster runes -> Once Ray Cluster context is finished and appwrapper is deleted, InstaScale deletes nodes. At no point will there be pending pods/services/routes etc. on your cluster

**Comment by tedhtchang** [comment]
@kevin85421 I need to put this PR on hold because mcad repo is under refactoring. The example yamls will not work any more.

**Comment by jbusche** [comment]
Hi @tedhtchang, I tested these steps using the kind cluster on my Mac M1 laptop, and it looked good!  I was able to install the items and test the appwrappers.  
```
oc get pods
NAME                                             READY   STATUS    RESTARTS   AGE
kuberay-operator-7cdbdb6f6d-d5tzl                1/1     Running   0          102m
raycluster-complete-1-head-qm98l                 1/1     Running   0          80m
raycluster-complete-1-worker-small-group-dx8sh   1/1     Running   0          80m
```
Nice work!

**Review by jbusche — APPROVED** [review]
LGTM

**Comment by tedhtchang** [comment]
@kevin85421 This doc should be good to merge.

**Comment by kevin85421** [comment]
@tedhtchang could you also update [README](https://github.com/ray-project/kuberay)?

**Review by kevin85421 — APPROVED** [review]
I do not retry the doc again after Sept, but some folks tried it and approved this PR.

File: README.md [hunk 11]
```diff
@@ -49,6 +49,7 @@ The KubeRay repository only contains documentation related to the development an
 * [Profiling with py-spy](https://docs.ray.io/en/master/cluster/kubernetes/k8s-ecosystem/pyspy.html#kuberay-pyspy-integration)
 * [KubeRay integration with Volcano](https://docs.ray.io/en/master/cluster/kubernetes/k8s-ecosystem/volcano.html#kuberay-volcano)
 * [Kubeflow: an interactive development solution](https://docs.ray.io/en/master/cluster/kubernetes/k8s-ecosystem/kubeflow.html#kuberay-kubeflow-integration)
+* [KubeRay integration with MCAD (Multi-Cluster-App-Dispatcher)](https://github.com/ray-project/kuberay/blob/master/docs/guidance/kuberay-with-MCAD.md)
```
Comment (kevin85421) [inline README.md:52]: Users who aren't familiar with MCAD might not understand what's the purpose of the doc. Maybe use something like "MCAD: An out-of-box Kubernetes Solution for Gang Scheduling and Multi-Cluster Job Dispatching". I am not familiar with MCAD, so you may be able to come up with a better title.

Comment (tedhtchang) [reply README.md:52]: How about this ?
MCAD: A Kubernetes controller for Queuing, Gang Scheduling jobs on single or multi-cluster environment.
